streams: Change UI for subscription status for "All streams."

raghavluthra20 opened the pull request:
<!-- Describe your pull request here.-->
Changes the UI for subscribing to streams to be more like the mobile version (zulip/zulip-mobile#5333).

- For unsubscribed streams, the on hover-checkmark is replaced by a "+" which is always displayed and has on-hover highlighting.
- For unsubscribed private streams, the "+" is disabled.
- Tooltips with appropriate messages are added on the "+" sign for the above 2 cases and for the disabled "Subscribe" button in case of unsubscribed private stream.


Fixes: #22217 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![light theme](https://user-images.githubusercontent.com/59819879/179309403-6fde033d-fdd0-4509-bd30-19b86b4d2a51.jpg)
![dark theme](https://user-images.githubusercontent.com/59819879/179309418-5b1e306c-9149-4235-a76f-385413934bf1.jpg)


https://user-images.githubusercontent.com/59819879/179309688-58dd5fb2-65f4-48c4-a7bb-40fa3281ee8a.mp4

https://user-images.githubusercontent.com/59819879/179309693-3aa0c07e-884f-4d1a-8da1-e542b30cef90.mp4

Link to discussion thread: https://chat.zulip.org/#narrow/stream/101-design/topic/Improve.20.22All.20Streams.22.20UI.20.2322217

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [ ] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

raghavluthra20 commented:
@alya I think the next step is to tweak the colors in the light and dark modes. Shall I start a discussion on #design?

sov-j commented:
Tested this PR on current Firefox on Linux. I checked 3 requirements listed in #22217 and I think the first one is implemented with reverse logic (this is also visible on the video in PR description). The rest works as expected.

The logic is IMHO reversed because on hover the "+" sign becomes faded (like inactive one) and is highlighted otherwise, while the intent was the opposite in #22217, per my understanding. Maybe this needs additional clarification?

raghavluthra20 commented:
> Maybe this needs additional clarification?

@sov-j Yes, I'm still waiting for clarification on the exact colors.

alya commented:
Agreed with @sov-j 's feedback -- hovering over the `+` should make it darker, not lighter. E.g. see "mute topic" buttons in Recent topics and the message header bar, etc.

alya commented:
> @alya I think the next step is to tweak the colors in the light and dark modes. Shall I start a discussion on #design?

Once you've made the above change, please go ahead and post in #design for feedback. You should include both screenshots and the technical specifications of the colors you used, and any motivation behind the choices you made.

alya commented:
In the future, there's no need to wait on my approval to start of a #design discussion. :)

timabbott commented:
@raghavluthra20 can you comment on whether you've resolved the color issue discussed above? I see you updated this yesterday, but wasn't sure if you'd addressed the feedback.

raghavluthra20 commented:
> @raghavluthra20 can you comment on whether you've resolved the color issue discussed above? I see you updated this yesterday, but wasn't sure if you'd addressed the feedback.

Yep! I've addressed the color issue. Further discussion is in this [thread](https://chat.zulip.org/#narrow/stream/101-design/topic/Improve.20.22All.20Streams.22.20UI.20.2322217) (also in the description).

alya commented:
Nice! A couple of small bits of feedback:

- For streams you can't join, let's try making the tooltip more succinct: "Cannot subscribe to private stream #foo" 
- The cursor when hovering over a disabled `+` should be the disabled button cursor, not a hand.

alya commented:
I was also noticing that sometimes I wasn't getting the hand cursor at all, just the regular mouse pointer. I'm not sure how to replicate...

raghavluthra20 commented:
@sahil839 I've addressed the 2 points above.
![Uploading a3{80cfd2c} (2942717f-dbd9-4875-af37-2fca1827e270 id=0x7f0a03c6 tag=BaseCommentFragment) …]()

timabbott commented on static/templates/stream_settings/browse_streams_list_item.hbs:
```diff
@@ -2,12 +2,22 @@
+        </div>
+    {{else}}
+        <div class="check sub_unsub_button {{#unless should_display_subscription_button}}disabled{{/unless}} tippy-zulip-tooltip"
+          data-tippy-content="{{#if should_display_subscription_button}}Subscribe to #{{name}}{{else}}Cannot subscribe to private stream #{{name}}.{{/if}} ">
```
The tooltips need to be tagged for translation. I think the best way to do that is with introducing a `template` element so we don't have to stuff the syntax into this inline attribute; there's one in `static/templates/right_sidebar.hbs`.

timabbott commented on static/styles/subscriptions.css:
```diff
@@ -562,6 +557,11 @@ h4.user_group_setting_subsection_title {
         fill: hsl(170, 48%, 54%);
     }
 
+    .disabled svg {
```
Might be better to write this as `.check.disabled svg`.

timabbott commented on static/styles/dark_theme.css:
```diff
@@ -581,6 +581,25 @@ body.dark-theme {
         background-color: hsla(0, 0%, 0%, 0.2);
     }
 
+    .stream-row {
```
Can you add a comment for this section explaining that this is how we implement coloring these SVGs.

raghavluthra20 replied:
I've added comments mentioning what each section is responsible for.

raghavluthra20 commented:
> I was also noticing that sometimes I wasn't getting the hand cursor at all, just the regular mouse pointer. I'm not sure how to replicate...

I tried with Chrome and Firefox but can't replicate.

alya commented:
I have a couple of pieces of feedback after playing around with the PR:

1. Let's add a tooltip when hovering over the checkmark: "Unsubscribe from #stream". 
2. I realized that while hovering over the `+` increases the contrast, hovering over the checkmark actually decreases the contrast, and it looks a bit odd.

One thing we can try is changing the checkmarks to match the "Send" button, both with and without hover. Could you try that and post screenshots, both light and dark theme?

Thanks for continuing to iterate on this!

raghavluthra20 commented:
Using the colors from the "Send" button.

![Screenshot from 2022-09-12 01-35-48](https://user-images.githubusercontent.com/59819879/189546968-337bd3ce-2746-4a58-a707-6e93c585e503.png)
![Screenshot from 2022-09-12 01-36-16](https://user-images.githubusercontent.com/59819879/189546969-4b2c4e91-ddac-42b5-a5ff-69f9133ac292.png)

I'll push the changes by tomorrow after some more testing.

alya commented:
Thanks for the update! Please post a comment next time you push an update that's ready for review, as otherwise it's easy to miss (like I did this time).

alya commented:
I found a  bugs with the tooltips when testing this (not the incorrect stream name):

<img width="555" alt="Screen Shot 2022-09-17 at 8 29 00 PM" src="https://user-images.githubusercontent.com/2090066/190884467-4e460064-50ec-45c2-a9e4-0d3b70390cff.png">

Please re-test and ping me for another review once it's been fixed. :)

alya commented:
In general, I'd encourage you to carefully test your changes to avoid extra review round trips. It'll save reviewers time and help your work get integrated more quickly.

raghavluthra20 commented:
@alya I have resolved wrong stream name bug. Adding a unique identifier (the name of the stream in this case) in the template id seems to fix it. Maybe we can try using something more concrete like an id instead of name.

raghavluthra20 commented:
I've also found a another small bug. Steps to reproduce:

1. Unsubscribe from a private stream.
2. Immediately after that, hover the mouse pointer over the disabled `subscribed` button. Notice the tool-tip doesn't appear.
3. Click on a different stream and then back to the previous private stream and hover over the button again. The tool-tip now starts appearing.


https://user-images.githubusercontent.com/59819879/191588268-006adcb9-c3e1-42eb-a3bc-c90de6410472.mp4

alya commented:
Good find! Do you have an idea of how to fix it? If not, perhaps you can get advice in the #frontend stream.

alya commented:
Hi @raghavluthra20 ! Have you been able to figure out a fix for the bug you found? I'd love to get this PR ready for code review so that we can integrate it.

raghavluthra20 commented:
> Hi @raghavluthra20 ! Have you been able to figure out a fix for the bug you found? I'd love to get this PR ready for code review so that we can integrate it.

Not yet. I started a [topic](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Tooltip.20not.20appearing.20immediately) in #frontend but there aren't any replies as of now.
I'll take a look next week to see if I can figure it out since I have college exams this week.

alya commented:
@m-e-l-u-h-a-n  or @sahil839 , perhaps you could help here? I'd like to get this PR integrated, and follow up by implementing the same UI in the new groups panel.

sahil839 commented:
The right way to add tooltip here would be using a class and adding the tooltip using that class in `tippyjs.js` and then add or remove the class while changing th state of the button.

The author of this PR seems to be inactive. @m-e-l-u-h-a-n are you available to complete this? Or else I can try doing it next week.

m-e-l-u-h-a-n commented:
Sure I'll take care of this.

timabbott commented:
Closing in favor of the more current https://github.com/zulip/zulip/pull/25578 for this issue.
